### PR TITLE
[Portal] Global Search

### PR DESCRIPTION
**Pull request by Noggling**
# Description

Please include a summary of the change and which issue is fixed.
List any external dependencies that are required for this change.

Completes: AB#FILL_IN_YOUR_ISSUE_ID

## Checklist:

-   [ ] I have performed a self-review of my own code.
-   [ ] I have linked my DevOps task using the AB# tag.
-   [ ] My code is easy to read, and comments are added where needed.
-   [ ] My code is covered by tests.

### CODE REVIEW

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    const navigate = useNavigate();
+
+    const { customHttpClient } = httpClient({
+        scope: 'api://195ed58a-9cb8-4d93-9e37-9ad315032baf/ReadWrite',
+    });
+    // const { procosysPlantId } = useFacility();
+    const { searchResult, search, registerSearchItem } = useGlobalSearch();
+
+    useEffect(() => {
+        const procosysSearchRegistration = registerSearchItem<SearchResponse>({
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:30]: move config to its own filte

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            },
+            searchMapper: (data) => {
+                if (!Array.isArray(data)) {
+                    return searchMapper(data);
+                }
+                return [];
+            },
+        });
+
+        const appsRegistration = registerSearchItem<AppSearchItem>({
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:54]: same as above

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            appsRegistration();
+        };
+    }, []);
+
+    useEffect(() => {
+        search(searchInput);
+    }, [searchInput]);
+
+    useEffect(() => {
+        window.onkeydown = (ev: KeyboardEvent) => {
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:87]: refactor to util function

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+    let index = 0;
+
+    return (
+        <Wrapper>
+            <ArrowNavigation>
+                <Search
+                    ref={searchRef}
+                    value={searchInput}
+                    placeholder={'Press F1 to search..'}
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:104]: change title on focus

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                    <SearchResult>
+                        <>
+                            {searchResult.map((searchType) => (
+                                <div key={searchType.type}>
+                                    <SearchResultHeading
+                                        typeTitle={searchType.title}
+                                        color={searchType.color}
+                                    />
+                                    {searchType.items.map((item) => {
+                                        index++;
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:119]: comment

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import { SearchResultHeading } from './SearchResultHeading';
+import { SearchResultItem } from './SearchResultItem';
+
+export const GlobalSearch = (): JSX.Element => {
+    const [searchInput, setSearchInput] = useState('');
+    const searchRef = useRef<HTMLInputElement>(null);
+    const { apps } = useRegistry();
+    const navigate = useNavigate();
+
+    const { customHttpClient } = httpClient({
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:23]: fix in infra

File: src/Core/GlobalSearh/Components/SearchResultItem.tsx [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    const {
+        selected,
+        focusProps: { ref, tabIndex, onClick },
+    } = useArrowNavigationWithFocusState(0, index);
+
+    function handleOnClick(): void {
+        action(id);
+        onClick();
+    }
+    const handleOnKeyPress: KeyboardEventHandler<HTMLDivElement> = (ev) => {
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/SearchResultItem.tsx:32]: event

File: src/Core/GlobalSearh/Components/SearchResultItemStyles.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { Typography } from '@equinor/eds-core-react';
+import { tokens } from '@equinor/eds-tokens';
+import styled from 'styled-components';
+
+export const Wrapper = styled.div<{ selected: boolean }>`
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/SearchResultItemStyles.ts:5]: isSelected

File: src/Core/GlobalSearh/Functions/searchApps.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        )
+        .slice(0, 5)
+        .map((app) => ({
+            key: app.shortName,
+            title: app.title,
+            description: app.tags.length > 0 ? `Tags: ${app.tags.toString()}` : '',
+            id: app.shortName,
+            uri: app.uri && app.uri(isProduction()),
+            group: app.groupe,
+        }));
```
Comment (Noggling) [inline src/Core/GlobalSearh/Functions/searchApps.ts:29]: move to mapper function

File: src/Core/GlobalSearh/Functions/searchMapper.ts [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                        objects: item.tag,
+                    };
+                },
+            });
+        }
+        return acc;
+    }, [] as SearchResult[]);
+}
+
+function searchPushItem(
```
Comment (Noggling) [inline src/Core/GlobalSearh/Functions/searchMapper.ts:95]: util function and in procosys config folder

File: src/Core/GlobalSearh/Types/ProcoSysSearch.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    totalCommPkgHits: number;
+    totalMcPkgHits: number;
+    totalTagHits: number;
+    totalPunchItemHits: number;
+    items: SearchItems[];
+    debugInfo: string;
+}
+
+export interface SearchItems {
+    key: string;
```
Comment (Noggling) [inline src/Core/GlobalSearh/Types/ProcoSysSearch.ts:13]: Move file to procosys config folder

File: src/apps/Handover/index.tsx [hunk 12]
```diff
@@ -34,6 +35,24 @@ export const handoverCreatorManifest = creator('SidesheetManifest');
 export const handoverCreatorComponent = creator('SidesheetComponentManifest');
 export const handoverResolverFunction = creator('ResolverFunction');
 
+const creatorComPkg = setupWorkspaceSidesheet<HandoverPackage, 'comPkg'>({
```
Comment (Noggling) [inline src/apps/Handover/index.tsx:38]: remove

File: src/apps/Tags/Components/Banner.tsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    align-items: center;
+`;
+
+interface TextItemProps {
+    title: string;
+    value?: string | number | JSX.Element;
+    onClick?: () => void;
+}
+
+export function TextItem({ title, value, onClick }: TextItemProps): JSX.Element {
```
Comment (Noggling) [inline src/apps/Tags/Components/Banner.tsx:29]: split file

File: src/apps/Tags/Components/Header.tsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    buttonContent = 'Open in ProCoSys',
+}: SidesheetHeaderContentProps) => {
+    return (
+        <Header>
+            {title}{' '}
+            <LinkContent target="_BLANK" href={url} rel="noreferrer">
+                <Button key="linkToProcosys" variant="ghost">
+                    {buttonContent}
+                </Button>
+                <Icon
```
Comment (Noggling) [inline src/apps/Tags/Components/Header.tsx:45]: smaler icon?

File: src/apps/widgets.ts [hunk 15]
```diff
@@ -39,7 +46,10 @@ const _widgets: WidgetManifest[] = [
     handoverCreatorManifest,
     workOrderCreatorManifest,
     releaseCreatorManifest,
+    comPkgManifest,
```
Comment (Noggling) [inline src/apps/widgets.ts:49]: remove

File: src/apps/widgets.ts [hunk 16]
```diff
@@ -54,6 +64,9 @@ const _widgetComponents: ComponentManifest[] = [
     workOrderCreatorComponent,
     workOrderCreatorComponent,
     releaseCreatorComponent,
+    comPkgComponent,
```
Comment (Noggling) [inline src/apps/widgets.ts:67]: remove

File: src/apps/functions.ts [hunk 17]
```diff
@@ -24,6 +24,9 @@ export const functions: FunctionManifest[] = [
     workOrderResolverFunction,
     mcResolverFunction,
     releaseResolverFunction,
+    comPkgResolverFunction,
```
Comment (Noggling) [inline src/apps/functions.ts:27]: remove

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx (hunk 6 above)
Comment (Noggling) [reply src/Core/GlobalSearh/Components/GlobalSearch.tsx:23]: and move to config

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+import { SearchResponse } from '../Types/ProcoSysSearch';
+import { SearchResult, Wrapper } from './GlobalSearchStyles';
+import { SearchResultHeading } from './SearchResultHeading';
+import { SearchResultItem } from './SearchResultItem';
+
+export const GlobalSearch = (): JSX.Element => {
+    const [searchInput, setSearchInput] = useState('');
+    const searchRef = useRef<HTMLInputElement>(null);
+    const { apps } = useRegistry();
+    const navigate = useNavigate();
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:21]: use other way of navigate, util function?

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+import { SearchItem } from '../Service/SearchApi';
+import { SearchResponse } from '../Types/ProcoSysSearch';
+import { SearchResult, Wrapper } from './GlobalSearchStyles';
+import { SearchResultHeading } from './SearchResultHeading';
+import { SearchResultItem } from './SearchResultItem';
+
+export const GlobalSearch = (): JSX.Element => {
+    const [searchInput, setSearchInput] = useState('');
+    const searchRef = useRef<HTMLInputElement>(null);
+    const { apps } = useRegistry();
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:20]: use readClientRegistry and move down

File: src/Core/GlobalSearh/Components/GlobalSearch.tsx [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+    const navigate = useNavigate();
+
+    const { customHttpClient } = httpClient({
+        scope: 'api://195ed58a-9cb8-4d93-9e37-9ad315032baf/ReadWrite',
+    });
+    // const { procosysPlantId } = useFacility();
```
Comment (Noggling) [inline src/Core/GlobalSearh/Components/GlobalSearch.tsx:26]: remove

File: src/Core/GlobalSearh/Config/ProCoSys/proCoSysSearchRequest.ts [hunk 21]
```diff
@@ -0,0 +1,25 @@
+import { getClientContext, httpClient } from '@equinor/lighthouse-portal-client';
+import { SearchResponse } from './types';
+
+export async function proCoSysSearchRequest(
+    searchText: string
+): Promise<SearchResponse | undefined> {
+    const { procosysPlantId } = getClientContext();
+    const { customHttpClient } = httpClient({
+        scope: 'api://195ed58a-9cb8-4d93-9e37-9ad315032baf/ReadWrite',
```
Comment (Gustav-Eikaas) [inline src/Core/GlobalSearh/Config/ProCoSys/proCoSysSearchRequest.ts:9]: Infra

File: src/Core/GlobalSearh/Service/SearchApi.ts [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+
+    private resetSearchResult = () => {
+        this.searchResults = [];
+
+        this.subscribers.forEach((subscriber) => {
+            subscriber.callBack([]);
+        });
+    };
+
+    private unsubscribeCreator = (subscriberId: number) => {
```
Comment (Gustav-Eikaas) [inline src/Core/GlobalSearh/Service/SearchApi.ts:142]: Could potentially do
 private unsubscribeCreator = (subscriberId: number) => {
 this.subscribers = this.subscribers.filter((s) => s["##id"] !== subscriberId)
 }
        But your approach is fine

File: src/apps/Tags/Functions/resolver.ts [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+    const { procosysPlantId } = getClientContext();
+    const { procosys } = httpClient();
+
+    const projectName = window.location.search;
+
+    const uri = `api/Tag/ByTagNos`;
+    const queryParameters =
+        `${projectName}&plantId=${procosysPlantId}&tagNos=${tagNo}&&api-version=4.1`.replace(
+            '#',
+            '%23'
```
[the remark this answers is not in the document]
Comment (Gustav-Eikaas) [reply src/apps/Tags/Functions/resolver.ts:14]: encodeURIComponent?

[the remark this answers is not in the document]
Comment (Noggling) [reply src/apps/Tags/Functions/resolver.ts:14]: funker ikke :/

**Review by Gustav-Eikaas — APPROVED** [review]

**Pull request merged**